tls-client/main.cpp

```diff
@@ -256,22 +263,32 @@ class HelloHTTPS {
 
 
         /* Read data out of the socket */
-        ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer, sizeof(_buffer));
+        offset = 0;
+        do {
+            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer + offset,
+                                   sizeof(_buffer) - offset - 1);
+            if (ret > 0)
+              offset += ret;
+
+            /* Check each of the flags */
+            _buffer[offset] = 0;
+            _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
+            _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+        } while ( (!_got200 || !_gothello) &&
+                (ret > 0 || ret == MBEDTLS_ERR_SSL_WANT_READ ||
+                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
```

Comment:
Since mbedtls lib expects and handles non blocking IO. The ssl_send and recv routines should not be blocking in terms of looping until data is received. Secondly in some cases like call to ```mbedtls_ssl_read```  the application does not know how much it want to receive and it sends buffer size instead of expected message size. Hence putting such loops in those routines can cause app to hang.